transport/src/main/java/io/netty/channel/CoalescingBufferQueue.java

```diff
@@ -0,0 +1,225 @@
+        if (bufAndListenerPairs.isEmpty()) {
+            return Unpooled.EMPTY_BUFFER;
+        }
+
+        ByteBuf toReturn = null;
+        int originalBytes = bytes;
+        for (;;) {
+            Object entry = bufAndListenerPairs.poll();
+            if (entry == null) {
+                break;
+            }
+            if (entry instanceof ChannelFutureListener) {
+                aggregatePromise.addListener((ChannelFutureListener) entry);
+                continue;
+            }
+            ByteBuf entryBuffer = (ByteBuf) entry;
+            if (entryBuffer.readableBytes() > bytes) {
+                // Add the buffer back to the queue as we can't consume all of it.
+                bufAndListenerPairs.addFirst(entryBuffer);
+                if (bytes == 0) {
+                    // If we can't consume any of it just exit.
+                    break;
+                } else {
+                    // Take a slice of what we can consume and retain it.
+                    entryBuffer = entryBuffer.readSlice(bytes).retain();
+                }
+            }
+            toReturn = compose(toReturn, entryBuffer);
+            bytes -= entryBuffer.readableBytes();
+            if (entry != entryBuffer) {
+                // We couldn't consume the full entry as we took a slice of it.
+                break;
+            }
+        }
+        readableBytes -= originalBytes - bytes;
+        assert readableBytes >= 0;
+        return toReturn;
+    }
+
+    /**
+     * Compose the current buffer with another.
+     */
+    private ByteBuf compose(ByteBuf current, ByteBuf next) {
+        if (current == null) {
+            return next;
+        } else if (current instanceof CompositeByteBuf) {
+            CompositeByteBuf composite = (CompositeByteBuf) current;
+            composite.addComponent(next);
+            composite.writerIndex(composite.writerIndex() + next.readableBytes());
+            return composite;
+        } else {
+            // Create a composite buffer to accumulate this pair and potentially all the buffers
+            // in the queue. Using +2 as we have already dequeued current and next.
+            CompositeByteBuf composite = alloc.compositeBuffer(bufAndListenerPairs.size() + 2);
+            composite.addComponent(current);
+            composite.addComponent(next);
+            return composite.writerIndex(current.readableBytes() + next.readableBytes());
+        }
+    }
+
+    /**
+     * The number of readable bytes.
+     */
+    public int readableBytes() {
+        return readableBytes;
+    }
+
+    /**
+     * Are there pending buffers in the queue.
+     */
+    public boolean isEmpty() {
+        return bufAndListenerPairs.isEmpty();
+    }
+
+    /**
+     *  Release all buffers in the queue and complete all listeners and promises.
+     */
+    public void releaseAndFailAll(Channel channel, Throwable cause) {
```

Comment:
I wonder if it would be better to pass the Channel in the constructor. As far as I know a queue should be only used by one Channel right ?  This would look nicer in terms of API  I think